### PR TITLE
Added dup2() and dup3() support for rust descriptors

### PR DESCRIPTION
**Pull request by stevenengler**

### CODE REVIEW

File: src/main/host/syscall/unistd.rs [hunk 1]
```diff
@@ -55,23 +55,86 @@ pub fn dup(ctx: &mut ThreadContext, args: &SysCallArgs) -> SyscallResult {
...
+    let old_fd = libc::c_int::from(args.get(0));
+    let new_fd = libc::c_int::from(args.get(1));
 
-    if let Some(flags) = flags {
-        new_desc.set_flags(flags);
+    // get the descriptor, or return early if it doesn't exist
+    let desc = match syscall::get_descriptor(ctx.process, old_fd)? {
+        CompatDescriptor::New(desc) => desc,
+        // we don't support dup2 for legacy descriptors
+        CompatDescriptor::Legacy(_) => return Err(nix::errno::Errno::ENOSYS.into()),
```
Comment (sporksmith) [inline src/main/host/syscall/unistd.rs:73]: Log a warning?

File: src/test/dup/test_dup.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    tests
+}
+
+fn test_dup() -> Result<(), String> {
+    let (read_fd, write_fd) = nix::unistd::pipe().unwrap();
+
+    test_utils::run_and_close_fds(&[write_fd, read_fd], || {
+        let write_fd_dup =
+            test_utils::check_system_call!(|| { unsafe { libc::dup(write_fd) } }, &[])?;
```
Comment (sporksmith) [inline src/test/dup/test_dup.rs:51]: Seems worth validating the end state a bit more. e.g. write to both the old and new descriptor and validate that data from both shows up on the read end? Same for other tests below.

**Review by sporksmith — APPROVED** [review]

File: src/main/host/syscall/unistd.rs (hunk 1 above)
Comment (stevenengler) [reply src/main/host/syscall/unistd.rs]: Good idea, added.

File: src/test/dup/test_dup.rs (hunk 2 above)
Comment (stevenengler) [reply src/test/dup/test_dup.rs]: The pipe tests already had this same test, but I moved it to the dup tests and run it for all dup(), dup2(), and dup3().

**Pull request merged**